### PR TITLE
PIM-5833: Added category filter on the product export

### PR DESCRIPTION
**Pull request by ClementGautier**
# Description
Add a category filter on the product export builder for CSV and XLSX profiles.

# Definition Of Done
| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | :white_check_mark:  
| Added Behats                      | :white_check_mark: 
| Changelog updated                 | :white_check_mark: 
| Review and 2 GTM                  | :ballot_box_with_check: 
| Micro Demo to the PO (Story only) | :white_check_mark: 
| Migration script                  | N/A
| Tech Doc                          | N/A

### CODE REVIEW

File: features/Pim/Behat/Context/JobContext.php [hunk 1]
```diff
@@ -31,6 +31,13 @@ public function theFollowingJobConfiguration($code, TableNode $table)
 
             if ($this->isJobParameterArray($property)) {
                 $value = $this->getMainContext()->listToArray($value);
+
+                if (in_array($property, ['categories_included', 'categories_excluded'])) {
```
Comment (juliensnz) [inline features/Pim/Behat/Context/JobContext.php]: we were only supposed to send checked categories

File: features/export/product-export-builder/export_products_by_categories.feature [hunk 2]
```diff
@@ -0,0 +1,76 @@
+@javascript
+Feature: Export products from any given categories
+  In order to use the enriched product data
+  As a product manager
+  I need to be able to export the products from any given categories
+
+#  default
```
Comment (juliensnz) [inline features/export/product-export-builder/export_products_by_categories.feature]: please add a little description before that. It seems that it's commented code

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/model/product-export/tree.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ * Defines a model that store included and excluded categories of a tree in order to perform queries like
+ * categories IN included AND categories NOT IN excluded.
+ *
+ * @author    Clement Gautier <clement.gautier@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+define(['underscore', 'backbone'], function (_, Backbone) {
+
+    return Backbone.Model.extend({
```
Comment (juliensnz) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/model/product-export/tree.js]: I don't think we should use backbone models anymore. We had issue with them in the past. Maybe discuss IRL with @phaseinducer, maybe it's OK

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    ],
+    function ($, _, Backbone, Routing, __, LoadingMask, template) {
+
+        return Backbone.View.extend({
+
+            template: _.template(template),
+            loadingNodes: 0,
+
+            events: {
+                'click #trees-list a': 'switchTree'
```
Comment (juliensnz) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: please, never us ids

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+
+                $tree.bind('check_node.jstree', function (e, d) {
+                    if ($(d.rslt.obj[0]).hasClass('jstree-root') === false) {
+                        var self = this;
+                        this.model.include(d.rslt.obj[0].id.replace('node_', ''));
+                        this.incrementTreeCount(d.inst.get_container().attr('id').replace('tree-', ''));
+
+                        // Open the node and check childrens if needed
+                        d.inst.open_node(d.rslt.obj, function () {
+                            $('#' + d.rslt.obj[0].id + '>ul>li').each(function () {
```
Comment (juliensnz) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: please add a comment explaining what this black magic does and add spaces around `>`

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+                }.bind(this));
+
+                $tree.bind('uncheck_node.jstree', function (e, d) {
+                    this.model.exclude(d.rslt.obj[0].id.replace('node_', ''));
+                    this.decrementTreeCount(d.inst.get_container().attr('id').replace('tree-', ''));
+                }.bind(this));
+
+                $tree.bind('load_node.jstree', function (e, d) {
+                    this.model.set('loading', this.model.get('loading') + 1);
+                    var self = this;
```
Comment (juliensnz) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: we are not in 2007 anymore, please use .bind(this)

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+                if (1 === count) {
+                    scope.find('.icon-ok').removeClass('green').addClass('gray');
+                }
+
+                scope.find('.badge').html(count - 1);
+            },
+
+            /**
+             * Render the tree in the element's HTML
```
Comment (juliensnz) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: HTML element

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js (hunk 6 above)
Comment (ClementGautier) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: I can't bind 'this' as I need the 'this' of the function scope where I use it. Unless it's possible to bind it as an other variable name. That's why I did it here and nowhere else...

File: features/Pim/Behat/Context/JobContext.php (hunk 1 above)
Comment (ClementGautier) [inline features/Pim/Behat/Context/JobContext.php]: I would prefer to explain you IRL why but, to make it short, it was not possible on specific cases to do it this way, even with a custom operator in the backend side. That's why we used a combination of 'IN CHILDREN' and 'NOT IN' to do the job

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js (hunk 5 above)
Comment (ClementGautier) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: already commented line 123

File: src/Pim/Bundle/ImportExportBundle/Form/Type/JobParameter/CategoriesType.php [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+     */
+    public function buildView(FormView $view, FormInterface $form, array $options)
+    {
+        $trees = $this->categoryRepository->getTrees();
+
+        $view->vars['trees'] = array_map(function (Category $category) {
+            return [
+                'id'    => $category->getId(),
+                'label' => $category->getLabel(),
+                'code'  => $category->getCode(),
```
Comment (ClementGautier) [inline src/Pim/Bundle/ImportExportBundle/Form/Type/JobParameter/CategoriesType.php]: can be removed

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/model/product-export/tree.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        /**
+         * Add a value to the excluded collection
+         *
+         * @param {string} id
+         */
+        exclude: function (id) {
+            this.set('included', _.without(this.get('included'), id.toString()));
+
+            if (!this.get('excluded').includes(id.toString())) {
+                this.set('excluded', this.get('excluded').concat(id.toString()));
```
Comment (ClementGautier) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/model/product-export/tree.js]: push instead of concat

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/model/product-export/tree.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        /**
+         * Add a value to the included collection
+         *
+         * @param {string} id
+         */
+        include: function (id) {
+            this.set('excluded', _.without(this.get('excluded'), id.toString()));
+
+            if (!this.get('included').includes(id.toString())) {
+                this.set('included', this.get('included').concat(id.toString()));
```
Comment (ClementGautier) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/model/product-export/tree.js]: same

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+                    }
+                }.bind(this));
+
+                $tree.bind('uncheck_node.jstree', function (e, d) {
+                    this.model.exclude(d.rslt.obj[0].id.replace('node_', ''));
+                    this.decrementTreeCount(d.inst.get_container().attr('id').replace('tree-', ''));
+                }.bind(this));
+
+                $tree.bind('load_node.jstree', function (e, d) {
+                    this.model.set('loading', this.model.get('loading') + 1);
```
Comment (ClementGautier) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: this.loadingNodes

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+            },
+
+            /**
+             * @param {string} treeId
+             */
+            initTree: function (treeId) {
+                var $tree = this.$('#tree-' + treeId);
+                $tree.jstree(this.config);
+
+                $tree.bind('check_node.jstree', function (e, d) {
```
Comment (ClementGautier) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: e => event, d => tree

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+                    if ($(d.rslt.obj[0]).hasClass('jstree-root') === false) {
+                        var self = this;
+                        this.model.include(d.rslt.obj[0].id.replace('node_', ''));
```
Comment (ClementGautier) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: set d.rslt.obj[0] into a variable named domNode

Comment (ClementGautier) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: use self.

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector.js [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+                    tree.remove();
+                });
+
+                modal.on('ok', function () {
+                    this.model.set('included', tree.model.get('included'));
+                    this.model.set('excluded', tree.model.get('excluded'));
+
+                    modal.close();
+                    modal.remove();
+                    tree.remove();
```
Comment (ClementGautier) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector.js]: remove the promise

File: features/export/product-export-builder/export_products_by_categories.feature [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    Then exported file of "csv_product_export" should contain:
+      """
+      sku;categories;enabled;family;groups
+      toys_games;toys_games;1;default;
+      dolls;dolls;1;default;
+      women;women;1;default;
+      clothing;clothing;1;default;
+      shoes;shoes;1;default;
+      jewelry;jewelry;1;default;
+      """
```
Comment (ClementGautier) [inline features/export/product-export-builder/export_products_by_categories.feature]: Add a behat that check the UI

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+        'oro/loading-mask',
+        'text!pim/template/product-export/categories-selector-tree',
+        'jquery.jstree'
+    ],
+    function ($, _, Backbone, Routing, __, LoadingMask, template) {
+
+        return Backbone.View.extend({
+
+            template: _.template(template),
+            loadingNodes: 0,
```
Comment (phaseinducer) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: Maybe a little comment would be useful about this as it's not straightforward.

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+
+            events: {
+                'click .tab-groups a': 'switchTree'
```
Comment (phaseinducer) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: Try not to use bootstrap classes to bind events, and more generally not display-related classes.

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+                    select_limit: 1,
+                    select_multiple_modifier: false
+                }
+            },
+
+            /**
+             * Overrides the constructor in order to enable data binding between model and view
+             */
+            initialize: function () {
+                this.model.set('loading', 0);
```
Comment (phaseinducer) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: Should be removed?

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/model/product-export/tree.js (hunk 3 above)
Comment (phaseinducer) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/model/product-export/tree.js]: Indeed but it's interesting in this case since we bind events to it.

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js (hunk 11 above)
Comment (phaseinducer) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: I don't know if this comment means that you remove the model and use a method instead. Finally using a model just for holding a value and displaying the loading mask is a bit overkill imho.

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+                    }
+                }.bind(this));
+            },
+
+            /**
+             * @param {jQuery.Event} event
+             */
+            switchTree: function (event) {
+                var $target = $(event.target);
+                var treeId = $target.closest('li').attr('id').replace('node_', '');
```
Comment (phaseinducer) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: fu, thank you jstree

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+
+                this.$('#trees-list').find('li').removeClass('active');
```
Comment (phaseinducer) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: `this.$('#trees-list li')` ;)

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+                this.$('#node_' + treeId).addClass('active');
+
+                this.$('#trees').find('> div').hide();
```
Comment (phaseinducer) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: idem

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 22]
```diff
@@ -0,0 +1,211 @@
...
+
+            /**
+             * @param {string} treeId
+             */
+            initTree: function (treeId) {
+                var $tree = this.$('#tree-' + treeId);
+                $tree.jstree(this.config);
+
+                $tree.bind('check_node.jstree', function (e, d) {
+                    if ($(d.rslt.obj[0]).hasClass('jstree-root') === false) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: yoda please

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 23]
```diff
@@ -0,0 +1,211 @@
...
+                        var self = this;
```
Comment (phaseinducer) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: Did you try to use `.bind()` on line 130?

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 24]
```diff
@@ -0,0 +1,211 @@
...
+                        }
+                    });
+
+                    this.model.set('loading', this.model.get('loading') - 1);
+
+                }.bind(this));
+            },
+
+            /**
+             * @param {string} id
```
Comment (phaseinducer) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: Please add a description

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 25]
```diff
@@ -0,0 +1,211 @@
...
+             */
+            incrementTreeCount: function (id) {
+                var scope = this.$('#node_' + id);
+                scope.find('.icon-ok').removeClass('gray').addClass('green');
+                scope.find('.badge').html(parseInt(scope.find('.badge').html(), 10) + 1);
+            },
+
+            /**
+             * @param {string} id
```
Comment (phaseinducer) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: idem

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 26]
```diff
@@ -0,0 +1,211 @@
...
+                $tree.bind('check_node.jstree', function (e, d) {
+                    if ($(d.rslt.obj[0]).hasClass('jstree-root') === false) {
+                        var self = this;
+                        this.model.include(d.rslt.obj[0].id.replace('node_', ''));
+                        this.incrementTreeCount(d.inst.get_container().attr('id').replace('tree-', ''));
+
+                        // Open the node and check direct childrens if needed
+                        d.inst.open_node(d.rslt.obj, function () {
+                            $('#' + d.rslt.obj[0].id + ' > ul > li').each(function () {
+                                if (!self.model.get('excluded').includes(this.id.replace('node_', '').toString())) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: Maybe create a var to improve readability?

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js (hunk 6 above)
Comment (phaseinducer) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: Strange because it shouldn't do any difference.

Comment (juliensnz) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: you never use this in this callback so I don't understand your answer :/

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js (hunk 13 above)
Comment (juliensnz) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: what mean "use self (evrywhere)" ?

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/ChannelController.php [hunk 27]
```diff
@@ -44,4 +44,18 @@ public function indexAction()
 
         return new JsonResponse($normalizedChannels);
     }
+
+    /**
+     * @return JsonResponse
+     */
+    public function getAction($identifier)
```
Comment (ClementGautier) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/ChannelController.php]: use indexAction

File: src/Pim/Component/Catalog/Normalizer/Structured/ChannelNormalizer.php [hunk 28]
```diff
@@ -91,15 +91,19 @@ protected function normalizeLocales(ChannelInterface $channel)
...
      *
      * @return string
      */
     protected function normalizeCategoryTree(ChannelInterface $channel)
     {
-        return $channel->getCategory()->getCode();
+        return [
+            'id'    => $channel->getCategory()->getId(),
+            'code'  => $channel->getCategory()->getCode(),
+            'label' => $channel->getCategory()->getLabel(),
```
Comment (ClementGautier) [inline src/Pim/Component/Catalog/Normalizer/Structured/ChannelNormalizer.php]: labels

File: features/Pim/Behat/Context/JobContext.php [hunk 29]
```diff
@@ -31,6 +31,13 @@ public function theFollowingJobConfiguration($code, TableNode $table)
 
             if ($this->isJobParameterArray($property)) {
                 $value = $this->getMainContext()->listToArray($value);
+
+                if (in_array($property, ['categories'])) {
```
Comment (ClementGautier) [inline features/Pim/Behat/Context/JobContext.php]: Ask our beloved product owner what she want: store code that can change in time (remove / recreate in another tree for example) or keep the ID storage.

File: features/Pim/Behat/Decorator/Tree/JsNodeDecorator.php [hunk 30]
```diff
@@ -57,6 +57,20 @@ public function isOpen()
     }
 
     /**
+     * Is the node checked ?
+     *
+     * @return boolean
+     */
+    public function isChecked()
```
Comment (ClementGautier) [inline features/Pim/Behat/Decorator/Tree/JsNodeDecorator.php]: isSelected and replace assertNull with a spin

File: features/export/product-export-builder/export_products_by_categories.feature [hunk 31]
```diff
@@ -0,0 +1,84 @@
...
+      toys_games;toys_games;1;default;
+      dolls;dolls;1;default;
+      """
+
+  @ce
+  Scenario: Export the products from a tree using the UI
+    When I am on the "csv_product_export" export job edit page
+    And I visit the "Content" tab
+    Then I should see the text "Categories All products"
+    When I press the "Edit" button
```
Comment (ClementGautier) [inline features/export/product-export-builder/export_products_by_categories.feature]: => ̀Edit categories` or `Select categories` ? ask our beloved UX

File: features/export/product-export-builder/export_products_by_categories.feature [hunk 32]
```diff
@@ -0,0 +1,84 @@
...
+      | action_figures         | action_figures         | default |
+      | dolls                  | dolls                  | default |
+      | video_games            | video_games            | default |
+      | clothing_shoes_jewelry | clothing_shoes_jewelry | default |
+      | women                  | women                  | default |
+      | clothing               | clothing               | default |
+      | shoes                  | shoes                  | default |
+      | jewelry                | jewelry                | default |
+      | men                    | men                    | default |
+    And the following jobs:
```
Comment (ClementGautier) [inline features/export/product-export-builder/export_products_by_categories.feature]: Override this method in the EE to be able to remove the `@ce`

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 33]
```diff
@@ -0,0 +1,223 @@
...
+                } else if (this.model.get('categories').includes(node[0].id.replace('node_', '').toString())) {
+                    data.inst.check_node(node);
+                }
+            },
+
+            /**
+             * Render the tree in the element's HTML when the channel category is fetched and bind events from jstree
+             */
+            render: function () {
+                this.trigger('increase_loading');
```
Comment (ClementGautier) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: load the LoadingMask manualy here

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 34]
```diff
@@ -0,0 +1,223 @@
...
+
+                FetcherRegistry.initialize().then(function () {
+                    FetcherRegistry.getFetcher('channel')
+                        .fetch(this.attributes.channel)
+                        .then(function (channel) {
+
+                            this.attributes.tree = channel.category;
```
Comment (ClementGautier) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: `this.currentTree` instead

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 35]
```diff
@@ -0,0 +1,223 @@
...
+
+                            this.$el.html(this.template({
+                                tree: this.attributes.tree,
+                                label: _.findWhere(
```
Comment (ClementGautier) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: use i18n module

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js [hunk 36]
```diff
@@ -0,0 +1,223 @@
...
+                                .on('uncheck_node.jstree', function (event, data) {
+                                    this.uncheckNode(data);
+                                }.bind(this))
+                                .on('load_node.jstree', function (event, data) {
+                                    this.loadNode(data);
+                                }.bind(this));
+
+                        }.bind(this))
+                        .done(function () {
+                            this.trigger('decrease_loading');
```
Comment (ClementGautier) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector-tree.js]: remove the loading mask here manualy

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector.js [hunk 37]
```diff
@@ -0,0 +1,112 @@
...
+        var TreeModal = Backbone.BootstrapModal.extend({
+            className: 'modal jstree-modal'
+        });
+
+        return Backbone.View.extend({
+
+            template: _.template(template),
+
+            events: {
+                'click button.btn.edit': 'open'
```
Comment (ClementGautier) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector.js]: remove button.btn

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector.js [hunk 38]
```diff
@@ -0,0 +1,112 @@
...
+            attributes: {
+                disabled: false
+            },
+
+            /**
+             * Overrides the constructor in order to enable data binding between model and view
+             */
+            initialize: function () {
+                this.model.bind('change:categories', function () {
+                    $('#' + this.$el.data('categories-input')).val(this.model.get('categories').join());
```
Comment (ClementGautier) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/product-export/categories-selector.js]: => join(', ')

File: src/Pim/Bundle/ImportExportBundle/Resources/public/templates/product-export/categories-selector-tree.html [hunk 39]
```diff
@@ -0,0 +1,3 @@
+<h4><%- label %></h4>
+<hr />
```
Comment (ClementGautier) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/templates/product-export/categories-selector-tree.html]: / ? not / ? that's the question. Ask our beloved UX

File: src/Pim/Bundle/UIBundle/Resources/public/css/pim.less [hunk 40]
```diff
@@ -1782,3 +1782,21 @@ h3.tab-header {
 .translation-container .validation-error {
   display: block;
 }
+
+.flex-space-between {
+  display:flex;
```
Comment (ClementGautier) [inline src/Pim/Bundle/UIBundle/Resources/public/css/pim.less]: space after `:`

File: src/Pim/Bundle/UIBundle/Resources/public/css/tree.less [hunk 41]
```diff
@@ -117,3 +117,26 @@
...
+
+.jstree li.separated {
+  margin-top: 20px;
+  border-top: solid 1px @color-purple_dark;
+}
+
+/** modal */
+
+.jstree-modal {
+
```
Comment (ClementGautier) [inline src/Pim/Bundle/UIBundle/Resources/public/css/tree.less]: no blank line

File: src/Pim/Component/Catalog/Normalizer/Structured/ChannelNormalizer.php [hunk 42]
```diff
@@ -91,15 +91,28 @@ protected function normalizeLocales(ChannelInterface $channel)
...
+            $normalizedTranslations[] = [
+                'locale' => $translation->getLocale(),
+                'label'  => $translation->getLabel(),
+            ];
+        }
+
+        return [
+            'id'           => $channel->getCategory()->getId(),
+            'code'         => $channel->getCategory()->getCode(),
+            'translations' => $normalizedTranslations,
```
Comment (ClementGautier) [inline src/Pim/Component/Catalog/Normalizer/Structured/ChannelNormalizer.php]: labels

File: src/Pim/Component/Connector/Reader/Database/ProductReader.php [hunk 43]
```diff
@@ -347,7 +347,31 @@ protected function getProductIdentifiersFilter(JobParameters $parameters)
...
+    /**
+     * Transform category fields into PQB filters
+     *
+     * @param JobParameters $parameters
+     *
+     * @return array
+     */
+    protected function getCategoryFilters(JobParameters $parameters)
+    {
+        $included = $parameters->get('categories');
```
Comment (ClementGautier) [inline src/Pim/Component/Connector/Reader/Database/ProductReader.php]: categories instead of included

File: src/Pim/Bundle/ImportExportBundle/Resources/public/js/model/product-export/tree.js (hunk 3 above)
Comment (ClementGautier) [inline src/Pim/Bundle/ImportExportBundle/Resources/public/js/model/product-export/tree.js]: Remove the `BackboneModel`

**Comment by ClementGautier** [comment]
Approved by peers during a MOB review
